### PR TITLE
feat(x509_cert): add proxy support

### LINKED ISSUE #9174: Add proxy support for inputs plugin x509_cert
## Feature Request

Add proxy support for inputs plugin x509_cert 

### Proposal:
add option **http_proxy**  as in http_response to support access a target via proxy

### Current behavior:
no proxy support

### Desired behavior:
proxy support is added

### Use case: 
Can't access a target site directly and need to configure a proxy

### PR DESCRIPTION
**Pull request by akrantz01**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.

resolves #9174

Adds proxy support for the `x509_cert` plugin. The supported proxy methods are HTTP CONNECT (TCP only) and SOCKS5.

~~**Note:** #9289 should be merged before this PR since all URLs are currently parsed as globs causing telegraf to fail to pull the certificate. Similarly, the `TestGatherTCPCert` test will fail because of this.~~ Rebased to master.

### CODE REVIEW

**Comment by mariuszjedrzejewski** [comment]
Is there any chances to resolve conficts and merge proxy support changes? Can't wait for this feature...

**Comment by sspaink** [comment]
@akrantz01 do you have time rebase this pr? Thanks!

**Comment by akrantz01** [comment]
Done!

**Comment by powersj** [comment]
I pushed a commit to resolve some conflicts post-README-revamp that was done. It required removing the description/sample-config consts and a test. Will wait on CI and then merge.

**Comment by powersj** [comment]
@akrantz01 I tried, but seem to have failed :) would you mind rebasing once more? sorry!

**Comment by akrantz01** [comment]
I'm a bit confused why `TestGatherTCPCert` needed to be removed. It seems like it would be helpful to keep that test.

**Comment by powersj** [comment]
> I'm a bit confused why `TestGatherTCPCert` needed to be removed. It seems like it would be helpful to keep that test.

You are right, I meant to only remove `TestStrings`

File: plugins/common/proxy/proxy.go
Comment (reimda) [inline plugins/common/proxy/proxy.go:45]: If the user doesn't set the proxy url, I would expect not to use a proxy at all, not to use the system proxy settings.

Does proxy.FromEnvironment() not use a proxy in this case or does it use the system proxy settings?

Maybe we need two settings: proxy_url for the url, and something like use_proxy (bool). If use_proxy is true but proxy_url isn't set, then use the system proxy.

**Review by reimda — CHANGES_REQUESTED** [review]

File: plugins/common/proxy/proxy.go
Comment (powersj) [reply plugins/common/proxy/proxy.go:45]: Having worked on a corporate network that required a proxy for some things, but not all, I prefer that applications only use the proxy when explicitly told to.

File: plugins/common/proxy/proxy.go
Comment (powersj) [inline plugins/common/proxy/proxy.go:12]: Can we rename `UseProxy` to `UseSystemProxy`? and `use_proxy` to `use_system_proxy`

File: plugins/common/proxy/proxy.go
Comment (powersj) [inline plugins/common/proxy/proxy.go:19]: @reimda, @sspaink and I were chatting and we believe this would break existing users who only have the proxy URL defined. In order to add this in a backwards-compatible way we need to implement a priority as such:

1. system proxy: if the `UseSystemProxy` is enabled, use the system proxy
2. HTTP proxy URL: if the `HTTPProxyURL` is set use this
3. no proxy

We may have users with configs today that only have the URL defined and not the boolean.

**Review by powersj — CHANGES_REQUESTED** [review]
@akrantz01 thank you for making the update so quickly! We have one more set of changes, but we know you are probably busy with schools work. If you would rather one of us pick this up please let us know! We really appreciate all the effort you always put into these PRs.

Thanks!

File: plugins/common/proxy/proxy.go
Comment (akrantz01) [reply plugins/common/proxy/proxy.go:19]: Sounds good. Should I make `TCPProxy` act the same way for consistency?

Comment (MyaLongmire) [reply plugins/common/proxy/proxy.go]: yes if you have the time :)

**Review by reimda — APPROVED** [review]
Thanks Alex!

**Comment by reimda** [comment]
@powersj I think Alex addressed all the concerns. Yours was the last review that requested changes. Could you take another look? If we all like this, we can resolve the conflicts one last time before merging.

**Comment by powersj** [comment]
Rebase + CI and ready to merge

**Pull request merged**
